lang/LangSource/PyrSlot.h

```diff
@@ -27,16 +27,398 @@ A PyrSlot is an 8-byte value which is either a double precision float or a
+    T value;
+#endif
+};
+
+static_assert(sizeof(PaddedNumber<uint32_t>) == 8);
+static_assert(sizeof(PaddedNumber<uint16_t>) == 8);
+static_assert(sizeof(PaddedNumber<uint8_t>) == 8);
+
+template <typename T> struct MaybePaddedPointer {
+    static_assert(std::is_pointer_v<T>);
+#if POINTER_NEEDS_PADDING && (BYTE_ORDER == BIG_ENDIAN)
+    int32_t padding;
+#endif
+
+    T ptr;
+
+    [[nodiscard]] T getPtr() const noexcept {
+#if POINTER_NEEDS_PADDING
+        return ptr; // If there is padding, the tag hasn't tainted the pointer, return it.
+#else
+        return reinterpret_cast<T>(reinterpret_cast<uintptr_t>(ptr) & pointerMask);
+#endif
+    }
+
+#if POINTER_NEEDS_PADDING && (BYTE_ORDER == LITTLE_ENDIAN)
+    int32_t padding;
+#endif
+};
+
+static_assert(sizeof(MaybePaddedPointer<void*>) == sizeof(double));
+
+// cpp reference
+template <class To, class From>
+std::enable_if_t<sizeof(To) == sizeof(From) && std::is_trivially_copyable_v<From> && std::is_trivially_copyable_v<To>,
+                 To>
+bit_cast(const From& src) noexcept {
+    static_assert(std::is_trivially_constructible_v<To>,
+                  "This implementation additionally requires "
+                  "destination type to be trivially constructible");
+    To dst;
+    std::memcpy(&dst, &src, sizeof(To));
+    return dst;
+}
+
+struct PyrNil {};
+
+union PyrSlot {
+private:
+    struct PrivateTag {};
+    constexpr PyrSlot(PrivateTag, uint64_t raw) noexcept: u_raw(raw) {}
+    constexpr PyrSlot(PrivateTag, double d) noexcept: u_double(d) {}
+
+    uint64_t u_raw;
+    double u_double;
+    PaddedNumber<char> u_char;
+    PaddedNumber<int32_t> u_int;
+    MaybePaddedPointer<void*> u_ptr;
+    MaybePaddedPointer<struct PyrObjectHdr*> u_objectHeader;
+    MaybePaddedPointer<struct PyrSymbol*> u_symbol;
+
+    template <uint64_t T> [[nodiscard]] bool tagChecker() const noexcept {
+        return isBoxed() && ((u_raw >> 48) == (T >> 48));
+    }
+    [[nodiscard]] bool isBoxed() const noexcept { return (u_raw & boxed) == boxed; }
+
+public:
+    constexpr PyrSlot() noexcept: u_raw(nilValue) {}
+    ~PyrSlot() noexcept = default;
+    PyrSlot(PyrSlot&&) noexcept = default;
+    PyrSlot(const PyrSlot&) noexcept = default;
+    PyrSlot& operator=(PyrSlot&&) noexcept = default;
+    PyrSlot& operator=(const PyrSlot&) noexcept = default;
+
+    friend bool operator==(PyrSlot lhs, PyrSlot rhs) noexcept {
+        if (lhs.isDouble()) {
+            // Doubles have odd comparison rules...
+            if (rhs.isDouble())
+                return lhs.getDouble() == rhs.getDouble();
+            return false;
+        }
+        if (rhs.isDouble())
+            return false;
+        return lhs.u_raw == rhs.u_raw;
+    }
+
+    [[nodiscard]] constexpr static PyrSlot make(double d) noexcept { return { PrivateTag(), d }; }
+    [[nodiscard]] static PyrSlot make(char c) noexcept {
+        return { PrivateTag(), charTag | static_cast<uint64_t>(bit_cast<uint8_t>(c)) };
```

Comment:
Minor but, I wonder if it would be better to switch to a three arguement private constructor, and then consolidate the `|` bit operation to one place?

Reply:
done